File: 1061-lexicographically-smallest-equivalent-string/1061-lexicographically-smallest-equivalent-string.cpp
class Solution {
public:
  int find(int repr[], int indx){
    if(repr[indx] == indx)
      return indx;
    return find(repr, repr[indx]);
  }
  void unite(int repr[], int i, int j){
    int a = find(repr, i);
    int b = find(repr, j);
    if(a < b)
      repr[b] = a;
    else
      repr[a] = b;
  }
  string smallestEquivalentString(string s1, string s2, string baseStr) {
    int repr[26];
    for(int i=0; i<26; i++)
      repr[i] = i;
    for(int i=0; i<s1.size(); i++){
//    for(int i=0; i<26; i++)
//      printf("(%d,%d), ", i, repr[i]);
//    cout<<'\n';
      char ch1 = s1[i], ch2 = s2[i];
      unite(repr, ch1-'a', ch2-'a');
    }
    for(int i=0; i<baseStr.size(); i++){
      char ch = baseStr[i]; 
      baseStr[i] = char('a' + find(repr, ch-'a'));
//      cout<<i<<'\n';
    }
    return baseStr;
  }
};

File: 1061-lexicographically-smallest-equivalent-string/1061-lexicographically-smallest-equivalent-string.py
class Solution:
  def smallestEquivalentString(self, s1: str, s2: str, baseStr: str) -> str:
    parent = list(range(26));
    def union(i, j):
      a = find(i)
      b = find(j)
      if(a < b):
        parent[b] = a
      else:
        parent[a] = b
    def find(i):
      if(parent[i] == i):
        return i;
      return find(parent[i]);
    for i in range(len(s1)):
      union(ord(s1[i])-ord('a'), ord(s2[i])-ord('a'))
    return ''.join([chr(find(ord(ch) - 97) + 97) for ch in baseStr])
      